androidApp/src/main/java/com/example/tasky/android/login/components/LoginTextField.kt

```diff
@@ -0,0 +1,137 @@
+package com.example.tasky.android.login.components
+
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material.icons.Icons
+import androidx.compose.material.icons.filled.Check
+import androidx.compose.material3.Icon
+import androidx.compose.material3.MaterialTheme.typography
+import androidx.compose.material3.OutlinedTextField
+import androidx.compose.material3.OutlinedTextFieldDefaults
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.input.PasswordVisualTransformation
+import androidx.compose.ui.text.input.VisualTransformation
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import com.example.tasky.android.R
+import com.example.tasky.android.theme.Light2
+import com.example.tasky.android.theme.LightBlue
+import com.example.tasky.android.theme.MyApplicationTheme
+import com.example.tasky.android.theme.Red
+
+@Composable
+fun LoginTextField(
+    text: String,
+    onTextChange: (String) -> Unit,
+    placeHolder: String,
+    validator: (String) -> Boolean,
+    errorText: String?,
+    isPassword: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    var showPassword by rememberSaveable {
+        mutableStateOf(false)
+    }
+
```

Comment:
I can think of a case where `remeberSaveable` is not what we want, since it possibly reveals sensitive information. Imagine the user has their password and then leaves the app, which then gets killed by the system. Coming back to this screen would immediately reveal their password since this boolean state is restored (they might have someone else watching their phone and wouldn’t expect this). Quite a rare scenario, but if it happens, it wouldn’t be that good in times where people use the same password across the internet 😅 

Thus, I'd strongly recommend to put all UI states in a ViewModel, since this will unburden your UI from handling state updates and rather let it focus on rendering the UI based on the state given to it.

Reply:
understand the point. but when we putting all these states to the ViewModel, would it increase a lot of boilerplate code?
for example, when we implement a form, there would be many similar textfields. If we keep this state in the custom TextField, we would not need to create a lot of similar states in the ViewModel. Is there any way to avoid this problem?

Reply:
I understand your concern, but I don’t see the boilerplate code 🤔 

If we have a screen with a bunch of textfields, would it not make sense that we have our state in the ViewModel? It is ultimately the place we want to use the data from those TextFields to either run validation, apply formatting, or use it to trigger a network call for a search query value.

Each piece of TextField state in your ViewModel serves a unique purpose to your feature (the form input screen) which needs to be modelled and the best place to store and manage that state is your ViewModel.

Reply:
It's not just for a clearer separation of responsibilities (ViewModel updates state, UI reads and interprets state), but also that the ViewModel would have more control over the state if we'd put it in there, since it would allow it to be observed, reactively combined with other possible states (e.g. to automatically validate the text field when the text changes, etc.)

If some of your states are in your UI and some are in the ViewModel, what is the rule behind when to put which where? Especially in a team this might not be that clear anymore and could lead to a bigger mess in the long run if everyone acts based on feeling 😄